lib/experimental/class-wp-navigation-gutenberg.php

```diff
@@ -0,0 +1,270 @@
+	 *
+	 * @return object WP_Term The classic navigation.
+	 */
+	public static function get_classic_menu_fallback() {
+		$classic_nav_menus = wp_get_nav_menus();
+
+		// If menus exist.
+		if ( $classic_nav_menus && ! is_wp_error( $classic_nav_menus ) ) {
+			// Handles simple use case where user has a classic menu and switches to a block theme.
+
+			// Returns the menu assigned to location `primary`.
+			$locations = get_nav_menu_locations();
+			if ( isset( $locations['primary'] ) ) {
+				$primary_menu = wp_get_nav_menu_object( $locations['primary'] );
+				if ( $primary_menu ) {
+					return $primary_menu;
+				}
+			}
+
+			// Returns a menu if `primary` is its slug.
+			foreach ( $classic_nav_menus as $classic_nav_menu ) {
+				if ( 'primary' === $classic_nav_menu->slug ) {
+					return $classic_nav_menu;
+				}
+			}
+
+			// Otherwise return the most recently created classic menu.
+			usort(
+				$classic_nav_menus,
+				function( $a, $b ) {
+					return $b->term_id - $a->term_id;
+				}
+			);
+			return $classic_nav_menus[0];
+		}
+	}
+
+	/**
+	 * Converts a classic navigation to blocks.
+	 *
+	 * @param  object $classic_nav_menu WP_Term The classic navigation object to convert.
+	 * @return array the normalized parsed blocks.
+	 */
+	public static function get_classic_menu_fallback_blocks( $classic_nav_menu ) {
+		// BEGIN: Code that already exists in wp_nav_menu().
+		$menu_items = wp_get_nav_menu_items( $classic_nav_menu->term_id, array( 'update_post_term_cache' => false ) );
+
+		// Set up the $menu_item variables.
+		_wp_menu_item_classes_by_context( $menu_items );
+
+		$sorted_menu_items = array();
+		foreach ( (array) $menu_items as $menu_item ) {
+			$sorted_menu_items[ $menu_item->menu_order ] = $menu_item;
+		}
+
+		unset( $menu_items, $menu_item );
+
+		// END: Code that already exists in wp_nav_menu().
+
+		$menu_items_by_parent_id = array();
+		foreach ( $sorted_menu_items as $menu_item ) {
+			$menu_items_by_parent_id[ $menu_item->menu_item_parent ][] = $menu_item;
+		}
+
+		$inner_blocks = static::parse_blocks_from_menu_items(
+			isset( $menu_items_by_parent_id[0] )
+			? $menu_items_by_parent_id[0]
+			: array(),
+			$menu_items_by_parent_id
+		);
+
+		return serialize_blocks( $inner_blocks );
+	}
+
+	/**
+	 * Turns menu item data into a nested array of parsed blocks
+	 *
+	 * @param array $menu_items               An array of menu items that represent
+	 *                                        an individual level of a menu.
+	 * @param array $menu_items_by_parent_id  An array keyed by the id of the
+	 *                                        parent menu where each element is an
+	 *                                        array of menu items that belong to
+	 *                                        that parent.
+	 * @return array An array of parsed block data.
+	 */
+	public static function parse_blocks_from_menu_items( $menu_items, $menu_items_by_parent_id ) {
```

Comment:
I wonder if this should not be a global function in wordpress? Looks like something other people may use for "reasons"?

Reply:
Do you mean this should be made into a global function?

Maybe. I was intending to pull out all the Classic menu specific conversion stuff into a separate class to encapsulate it.

Reply:
I don't think we should encapsulate in a private way the function that converts menus data to block data, seems like something anyone should be able to do.

Reply:
No I just mean in terms of colocating related code and avoiding colocating unrelated code 😄

Reply:
I've now extracted a dedicated class.